### PR TITLE
e2e: add client tests

### PR DESCRIPTION
**Pull request by hexfusion**
This PR adds basic e2e testing for the api client to ensure no regressions in logic. More robust tests will be added once #85 merges.

Signed-off-by: Sam Batschelet <sam.batschelet@avalabs.org>

### CODE REVIEW

File: spacesvm/src/api/client.rs [hunk 1]
```diff
@@ -23,13 +30,19 @@ pub use http::Uri;
...
     id: u64,
     client: HyperClient<C>,
-    pub uri: Uri,
+    endpoint: Uri,
+    private_key: Option<Key>,
 }
 
 impl Client<HttpConnector> {
-    pub fn new(uri: Uri) -> Self {
+    pub fn new(endpoint: Uri) -> Self {
```
Comment (hexfusion) [inline spacesvm/src/api/client.rs:38]: lets make this a &str and then validate as Uri

File: spacesvm/src/api/client.rs [hunk 2]
```diff
@@ -40,6 +53,15 @@ impl Client<HttpConnector> {
         Id::Num(id)
     }
 
+    pub fn set_endpoint(&mut self, endpoint: Uri) {
+        self.endpoint = endpoint;
+    }
+
+    pub fn set_private_key(mut self, private_key: Key) -> Self {
```
Comment (hexfusion) [inline spacesvm/src/api/client.rs:60]: this feels sort of awkward to have a single builder method.

File: spacesvm/src/api/client.rs (hunk 1 above)
Comment (hexfusion) [reply spacesvm/src/api/client.rs:44]: maybe down the road client has a builder but for not I think how it is works fine.

**Comment by hexfusion** [comment]
tungstenite failure now resolved https://github.com/snapview/tungstenite-rs/issues/321

File: spacesvm/src/api/client.rs [hunk 3]
```diff
@@ -1,48 +1,80 @@
...
 use hyper::{body, client::HttpConnector, Body, Client as HyperClient};
-use jsonrpc_core::{Call, Id, MethodCall, Params, Version};
+use jsonrpc_core::{Call, Id, MethodCall, Params, Value, Version};
 use serde::de;
 
 pub use http::Uri;
+use tokio::sync::RwLock;
 
-/// HTTP client for interacting with the API, assumes single threaded use.
+/// HTTP client for interacting with the API.
```
Comment (exdx) [inline spacesvm/src/api/client.rs:31]: nit: The `Arc` makes it obvious but looks like the Client supports multi-threaded use now. Worth adding a comment?

File: tests/e2e/src/tests/mod.rs [hunk 4]
```diff
@@ -177,6 +181,69 @@ async fn e2e() {
...
+        .await
+        .expect("decodeTx success");
+    log::info!("decode set response from {}: {:?}", chain_url, resp);
+
+    log::info!("issue set tx request...");
+    let resp = scli
+        .issue_tx(&resp.typed_data)
+        .await
+        .expect("issue_tx success");
+    log::info!("issue tx tx response from {}: {:?}", chain_url, resp);
```
Comment (hexfusion) [inline tests/e2e/src/tests/mod.rs:241]: echo :)

File: spacesvm/src/api/client.rs [hunk 5]
```diff
@@ -1,48 +1,80 @@
...
+
+    pub async fn set_private_key(&self, private_key: Key) {
+        let mut inner = self.inner.write().await;
+        inner.private_key = Some(private_key);
+    }
+
     /// Returns a serialized json request as string and the request id.
-    pub fn raw_request(&mut self, method: &str, params: &Params) -> (Id, String) {
-        let id = self.next_id();
+    pub async fn raw_request(&self, method: &str, params: &Params) -> (Id, String) {
```
Comment (exdx) [inline spacesvm/src/api/client.rs:76]: Is there any way this function could fail? Maybe returning a `Result` is safer?

**Review by exdx — APPROVED** [review]

**Pull request merged**